### ISSUE TITLE
assign operator work incorrectly in travis

### ISSUE THREAD

**Issue by sangyx** [arrayfire/arrayfire#2756]
I run the code in my machine correct. But it can't work currectly in travis.
code:
```cpp
inline std::tuple<af::array, af::array, af::array, af::array> train_test_split(af::array & X, af::array & y, float train_size, int random_state=-1, bool shuffle=true)
{
	int m = X.dims(0), n = X.dims(1), p = y.dims(1);
	int train_num = ceil(m * train_size), test_num = m - train_num;
	std::vector<int> idx(m);
	std::iota(idx.begin(), idx.end(), 0);

	if(shuffle)
		std::shuffle(idx.begin(), idx.end(), std::default_random_engine(random_state));

	af::array X_train(train_num, n), y_train(train_num, p), X_test(test_num, n), y_test(test_num, p);
	for(int i = 0; i < train_num; ++i)
	{
		X_train.row(i) = X.row(idx[i]);
		y_train.row(i) = y.row(idx[i]);
		af_print(X_train.row(i));
		af_print(X.row(idx[i]));
	}

	for(int i = train_num; i < m; ++i)
	{
		X_test.row(i-train_num) = X.row(idx[i]);
		y_test.row(i-train_num) = y.row(idx[i]);
	}

	return std::make_tuple(X_train, y_train, X_test, y_test);
}
``` 

the output is:
```
X_train.row(i)
[1 2 1 1]
    0.0000     0.0000 
X.row(idx[i])
[1 2 1 1]
    1.1966     4.9519 
X_train.row(i)
[1 2 1 1]
    1.6953     0.0000 
X.row(idx[i])
[1 2 1 1]
    1.1768     3.1670 
X_train.row(i)
[1 2 1 1]
    0.0000     0.0000 
X.row(idx[i])
[1 2 1 1]
   -1.0766    -3.1819 
```

.travis.yml
```
language: cpp

env:
  global:
    # Ubuntu version
    - LINUX_DIST=bionic
    - DEPS_DIR=${TRAVIS_BUILD_DIR}/deps

matrix:
  include:
    - os: linux
      dist: bionic
      sudo: true
      compiler: gcc
      addons:
        apt:
          sources:
            - ubuntu-toolchain-r-test
          packages: &precise_latest_boost_packages
            - gcc-7
            - g++-7
            - build-essential
            - libfreeimage-dev
            - cmake-curses-gui
            - libopenblas-dev
            - libfftw3-dev
            - liblapacke-dev
            - libboost-all-dev
            - lcov
            - libgtest-dev
            - cmake

# before_install:
#     # Combine global build options with OS/compiler-dependent options
#     - export CMAKE_OPTIONS=${CMAKE_OPTIONS}" "${ENV_CMAKE_OPTIONS}
#     - export CXX_FLAGS=${CXX_FLAGS}" "${ENV_CXX_FLAGS}

install:
    - echo ${PATH}
    - echo ${CXX}
    - ${CXX} --version
    - ${CXX} -v
    - sudo pip install codecov

    - mkdir -p ${DEPS_DIR}/arrayfire
    - wget https://arrayfire.s3.amazonaws.com/3.6.4/ArrayFire-v3.6.4_Linux_x86_64.sh -O /${DEPS_DIR}/ArrayFire-v3.6.4_Linux_x86_64.sh
    - sudo sh ${DEPS_DIR}/ArrayFire-v3.6.4_Linux_x86_64.sh --include-subdir --prefix=/opt
    - sudo sh -c "echo /opt/arrayfire/lib64 > /etc/ld.so.conf.d/arrayfire.conf"
    - sudo ldconfig

    - cd /usr/src/gtest
    - sudo mkdir build
    - cd build/
    - sudo cmake ..
    - sudo make
    - sudo cp libgtest*.a /usr/local/lib
    - cd ${TRAVIS_BUILD_DIR}

script:
    ############################################################################
    # Build main and tests
    ############################################################################
      - sudo mkdir -p build
      - sudo mkdir -p bin
      - cd build
      - sudo cmake -DCMAKE_BUILD_TYPE=Coverage -DBUILD_TESTS=ON -DBUILD_CUDA=OFF ..
      - sudo make MLKIT_TEST_COV
```

**Comment by umar456** [comment]
I am not sure I understand. Is the code giving incorrect results or you can't compile on Travis? Could you give me more information about the result or the error you are experiencing.

**Comment by sangyx** [comment]
For example, I want to assign `X.row(0)` (= [1.1966     4.9519]) to `X_train.row(1)`. But the code `X_train.row(1) = X.row(0);` doesn't work correctly. After assign, the X_train.row(1) = [1.6953     0.0000] is a random value.

**Comment by umar456** [comment]
I am unable to reproduce this on my machine. I tested using CPU and CUDA backends. Here is the code:

```
        array a = randu(10, 2);
        array b = randu(2, 2);

        af_print(a);
        af_print(b);

        a.row(3) = b.row(1);

        af_print(a);
```

Can you test the above code and tell me if you are seeing the same issue?

**Comment by sangyx** [comment]
I run it correctly in my machine. But the same issue happened in the travis. The output is:
···
a
[10 2 1 1]
    0.6010     0.1583 
    0.0278     0.3712 
    0.9806     0.3543 
    0.2126     0.6450 
    0.0655     0.9675 
    0.5497     0.3636 
    0.2864     0.4165 
    0.3410     0.5814 
    0.7509     0.8962 
    0.4105     0.3712 
b
[2 2 1 1]
    0.6755     0.5232 
    0.6105     0.5567 
a
[10 2 1 1]
    0.6010     0.1583 
    0.0278     0.3712 
    0.9806     0.3543 
    0.2126     0.6450 
    0.0655     0.9675 
    0.5497     0.3636 
    0.2864     0.4165 
    0.3410     0.5814 
    0.7509     0.8962 
    0.4105     0.3712 
···
It seems like the assign operator doesn't work in the travis.

**Comment by umar456** [comment]
This is strange. Can you try to build ArrayFire on travis and use that instead of our official installers?

**Comment by sangyx** [comment]
I build ArrayFire on travis and the problem solved.

**Comment by 9prady9** [comment]
> I am unable to reproduce this on my machine. I tested using CPU and CUDA backends. Here is the code:
> 
> ```
>         array a = randu(10, 2);
>         array b = randu(2, 2);
> 
>         af_print(a);
>         af_print(b);
> 
>         a.row(3) = b.row(1);
> 
>         af_print(a);
> ```
> 
> Can you test the above code and tell me if you are seeing the same issue?

@sangyx Using this code snippet, I was able to reproduce the problem with 3.6.4 installer, however the problem was fixed in 3.7.* already.
Here is the output using 3.7.2
```
./helloworld/helloworld_cpu 
ArrayFire v3.7.2 (CPU, 64-bit Linux, build 218dd2c)
[0] AMD: AMD A10-7850K Radeon R7, 12 Compute Cores 4C+8Ga
[10 2 1 1]
    0.6010     0.1583 
    0.0278     0.3712 
    0.9806     0.3543 
    0.2126     0.6450 
    0.0655     0.9675 
    0.5497     0.3636 
    0.2864     0.4165 
    0.3410     0.5814 
    0.7509     0.8962 
    0.4105     0.3712 

b
[2 2 1 1]
    0.6755     0.5232 
    0.6105     0.5567 

a
[10 2 1 1]
    0.6010     0.1583 
    0.0278     0.3712 
    0.9806     0.3543 
    0.6105     0.5567 
    0.0655     0.9675 
    0.5497     0.3636 
    0.2864     0.4165 
    0.3410     0.5814 
    0.7509     0.8962 
    0.4105     0.3712 
```
Using 3.6.4 the output doesn't change at all.
```
./helloworld/helloworld_cpu 
ArrayFire v3.6.4 (CPU, 64-bit Linux, build 1b8030c)
[0] AMD: AMD A10-7850K Radeon R7, 12 Compute Cores 4C+8G, 31054 MB, Max threads(4) 
a
[10 2 1 1]
    0.6010     0.1583 
    0.0278     0.3712 
    0.9806     0.3543 
    0.2126     0.6450 
    0.0655     0.9675 
    0.5497     0.3636 
    0.2864     0.4165 
    0.3410     0.5814 
    0.7509     0.8962 
    0.4105     0.3712 

b
[2 2 1 1]
    0.6755     0.5232 
    0.6105     0.5567 

a
[10 2 1 1]
    0.6010     0.1583 
    0.0278     0.3712 
    0.9806     0.3543 
    0.2126     0.6450 
    0.0655     0.9675 
    0.5497     0.3636 
    0.2864     0.4165 
    0.3410     0.5814 
    0.7509     0.8962 
    0.4105     0.3712 
```

I think the below is the reason behind the behavior in 3.6.4

The `#if` guards that encapsulate the `operator=` signatures in our `af/array.h` header are incomplete/incorrect. These were corrected in 3.7 with appropriate guards that take into account different compilers and versions on all operation systems. coincidentally we never encountered a problem until it showed up in the travis setup.

Thanks once again for reporting it up with us. This should be fixed in latest stable release and development branch.
